compiler/src/dotty/tools/dotc/transform/init/ParamOverridingCheck.scala

```diff
@@ -0,0 +1,343 @@
+        overridingSym = paramSym.overridingSymbol(overridingClass.asClass) if overridingSym.exists
+      do
+        // println(paramSym.show + " is overridden in " + thisRef.klass)
+        if !overridingSym.is(Flags.ParamAccessor) then
+          report.warning("Overriding parameter " + paramSym.show + " in " + klass.show + " as non-parameter in " + overridingSym.owner.show, overridingSym.defTree)
+        else
+          val overridingValue = thisRef.field(overridingSym)
+          if thisRef.field(paramSym) != overridingValue then
+            // println(overridingSym.show + " in " + thisRef.klass + " = " + overridingValue)
+            // println(paramSym.show + " in " + klass + " = " + thisRef.field(paramSym))
+            report.warning("Incorrect overriding: " + paramSym.show + " in " + klass.show + " has a different value from " + overridingSym.show + " in " + overridingSym.owner.show, overridingSym.defTree)
+
+  // ---------------------- Semantic definition --------------------------------
+
+  /** Evaluate a list of expressions */
+  def eval(exprs: List[Tree], thisV: ThisRef, klass: ClassSymbol): Contextual[List[Value]] =
+    exprs.map { expr => eval(expr, thisV, klass) }
+
+  /** Evaluate arguments of methods */
+  def evalArgs(args: List[Arg], thisV: ThisRef, klass: ClassSymbol): Contextual[List[Value]] =
+    for arg <- args yield eval(arg.tree, thisV, klass)
+
+  /** Evaluate an expression with the given value for `this` in a given class `klass`
+   *
+   *  Note that `klass` might be a super class of the object referred by `thisV`.
+   *  The parameter `klass` is needed for `this` resolution. Consider the following code:
+   *
+   *  class A {
+   *    A.this
+   *    class B extends A { A.this }
+   *  }
+   *
+   *  As can be seen above, the meaning of the expression `A.this` depends on where
+   *  it is located.
+   *
+   */
+  def eval(expr: Tree, thisV: ThisRef, klass: ClassSymbol): Contextual[Value] =
+    expr match
+      case Ident(nme.WILDCARD) =>
+        Skolem(expr)
+
+      case id @ Ident(name) if !id.symbol.is(Flags.Method)  =>
+        assert(name.isTermName, "type trees should not reach here")
+        eval(expr.tpe, thisV, klass, expr)
+
+      case Select(qualifier, name) =>
+        val qual = eval(qualifier, thisV, klass)
+
+        name match
+          case OuterSelectName(_, _) =>
+            Skolem(expr)
+          case _ =>
+            qual.select(expr.symbol, receiver = qualifier.tpe, source = expr)
+
+      case _: This =>
+        eval(expr.tpe, thisV, klass, expr)
+
+      case Typed(expr, tpt) =>
+        eval(expr, thisV, klass)
+
+      case NamedArg(name, arg) =>
+        eval(arg, thisV, klass)
+
+      case Literal(_) =>
+        Skolem(expr)
+
+      case _ =>
+        Skolem(expr)
+
+  /** Handle semantics of leaf nodes */
+  def eval(tp: Type, thisV: ThisRef, klass: ClassSymbol, source: Tree): Contextual[Value] = log("evaluating " + tp.show, printer) {
```

Comment:
In comments, explain for each parameter which specific type/thisref/class/tree it represents.

The same suggestion applies to the other `eval` above.

It would be good to apply the same suggestion to the init checker as well (but in a separate PR). There, often a method will take several klasses as parameters, for example, and it's not immediately clear which one refers to which class in the context of what is being checked.